server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -481,8 +440,19 @@ private void createIndicesAndUpdateShards(final ClusterState state) {
         final Map<Index, List<ShardRouting>> indicesToCreate = new HashMap<>();
         for (ShardRouting shardRouting : localRoutingNode) {
             if (failedShardsCache.containsKey(shardRouting.shardId()) == false) {
+                ShardId shardId = shardRouting.shardId();
                 final Index index = shardRouting.index();
                 final var indexService = indicesService.indexService(index);
+                if (shardRouting.initializing() == false && (indexService == null || indexService.getShardOrNull(shardId.id()) == null)) {
+                    // the master thinks we are active, but we don't have this shard at all, mark it as failed
+                    sendFailShard(
+                        shardRouting,
+                        "master marked shard as active, but shard has not been created, mark shard as failed",
+                        null,
+                        state
+                    );
+                    continue;
+                }
```

Comment:
nit: bit odd to `continue` past an `if/else`...

```suggestion
                } else
```